# Import Libraries

In [17]:
import pandas as pd 
pd.options.plotting.backend = "plotly"
import numpy as np

# Load Datasets

In [36]:
data_train = pd.read_parquet("../data/interim/data_train.parquet")

In [37]:
for column in data_train:
    print(f'{column} = {(data_train[column].isnull().sum()/len(data_train))*100}')

Efficiency = 10.354532306609858
BoilTime = 10.328473335157462
Color = 10.331404969445856
BoilGravity = 13.758485452579187
OG = 10.235638249358297
Name = 10.364630158047662
Size(L) = 10.216419757912156
PrimaryTemp = 35.62098528971062
BrewMethod = 10.279287026541063
SugarScale = 10.315443849431263
ABV = 10.310883529427093
FG = 10.32261006658067
IBU = 10.29622535798512
Style = 10.893627278531316


# Description and Data Cleaning

## Variable definitions
- Input/Output variables
- Variable Types (Categorical or Numeric)
- Data Types (int, float, string, factor, boolean, ...)

|Feature Name| Type | Description and values | % Missing
|--|--|--|--|
| Name | Categorical (nominal) | The beer name | 10.30% 
| Size(L)| Number | Amount brewed for recipe listed | 10.23% 
| OG | Number | Specific gravity of wort before fermentation | 10.25% 
| FG | Number (nominal) | Specific gravity of wort after fermentation | 10.30% 
| ABV | Number | Alcohol By Volume | 10.29% 
| IBU | Number | International Bittering Units | 10.28% 
| Color | Number | Standard Reference Method - light to dark ex. 40 = black-  | 10.32% 
| BoilTime | Number | Time wort is boiled | 10.38% 
| BoilGravity | Number | Specific gravity of wort before the boil | 13.76% 
| Efficiency | Number | Beer mash extraction efficiency - extracting sugars from the grain during mash | 10.36% 
| SugarScale | Categorical (nominal) | Scale to determine the concentration of dissolved solids in wort | 10.30%
| BrewMethod | Categorical (nominal) | Various techniques for brewing | 10.24% 
| PrimaryTemp | Number | Temperature at the fermenting stage | 35.63% 
| Style| Categorical (nominal) | The type of brew | 10.87% 



In [38]:
data_train.sample(10)

Efficiency      BoilTime         Color   BoilGravity       OG  \
121785        75.0  6.000000e+01  2.878000e+01  1.066000e+03  1076.00   
252861        65.0  6.000000e+01  5.670000e+00           NaN  1066.00   
33041         75.0  9.000000e+01           NaN  1.055000e+03      NaN   
105331        35.0 -3.541287e+11 -3.541287e+11  1.120000e+00  1065.00   
203407        70.0  6.000000e+01  3.970000e+00  1.037000e+03  1046.00   
189174        70.0  7.500000e+01  1.398000e+01  1.044000e+03     1.06   
292439        70.0  6.000000e+01  7.970000e+00  1.054000e+03  1066.00   
175208        76.0  6.000000e+01  5.550000e+00  1.045000e+03  1055.00   
295536        70.0  9.000000e+01  2.182000e+01  1.049000e+03  1075.00   
143662        72.0  6.000000e+01  2.896000e+01 -3.541287e+11  1093.00   

                         Name       Size(L)   PrimaryTemp     BrewMethod  \
121785         Cool Calm Peat  3.500000e+01  1.800000e+01      All Grain   
252861  Bells two hearted ipa  4.542000e+01  1.889000e+01           BIAB   
33041                     NaN  5.680000e+00           NaN            NaN   
105331           White Zombie  2.082000e+01  1.833000e+01  -354128741541   
203407        Dirty Wheat Boy  2.271000e+01  2.000000e+01      All Grain   
189174           M.I.A. Amber  2.082000e+01  1.833000e+01      All Grain   
292439         Maybe Pale Ale  4.500000e+01  1.700000e+01      All Grain   
175208      Motueka S.M.A.S.H  2.150000e+01 -3.541287e+11      All Grain   
295536         Belgian dubbel  2.082000e+01  8.568422e+11      All Grain   
143662      Pukin pumpkin ale -3.541287e+11  2.111000e+01      All Grain   

              SugarScale           ABV       FG           IBU  \
121785  Specific Gravity  8.568422e+11  1017.00  3.962000e+01   
252861               NaN  6.720000e+00      NaN  6.558000e+01   
33041   Specific Gravity  1.013000e+01      NaN  5.966000e+01   
105331  Specific Gravity  6.420000e+00  1016.00  3.862000e+01   
203407  Specific Gravity  5.300000e+00  1005.00  1.523000e+01   
189174  Specific Gravity  5.880000e+00  1015.00  2.759000e+01   
292439  Specific Gravity  6.020000e+00     1.02  3.983000e+01   
175208  Specific Gravity  5.730000e+00  1011.00  2.978000e+01   
295536  Specific Gravity  7.430000e+00  1019.00  8.568422e+11   
143662  Specific Gravity  8.740000e+00  1026.00  7.120000e+01   

                         Style  
121785           Baltic Porter  
252861                     NaN  
33041   Specialty IPA: Rye IPA  
105331              Blonde Ale  
203407     American Wheat Beer  
189174                     NaN  
292439       American Pale Ale  
175208    American Light Lager  
295536          Belgian Dubbel  
143662          Specialty Beer

In [39]:
data_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 306996 entries, 0 to 306995
Data columns (total 14 columns):
 #   Column       Non-Null Count   Dtype   
---  ------       --------------   -----   
 0   Efficiency   275208 non-null  float64 
 1   BoilTime     275288 non-null  float64 
 2   Color        275279 non-null  float64 
 3   BoilGravity  264758 non-null  float64 
 4   OG           275573 non-null  float64 
 5   Name         275177 non-null  category
 6   Size(L)      275632 non-null  float64 
 7   PrimaryTemp  197641 non-null  float64 
 8   BrewMethod   275439 non-null  category
 9   SugarScale   275328 non-null  category
 10  ABV          275342 non-null  float64 
 11  FG           275306 non-null  float64 
 12  IBU          275387 non-null  float64 
 13  Style        273553 non-null  category
dtypes: category(4), float64(10)
memory usage: 28.2 MB


Null Values

In [40]:
data_train.isnull().sum()

Efficiency      31788
BoilTime        31708
Color           31717
BoilGravity     42238
OG              31423
Name            31819
Size(L)         31364
PrimaryTemp    109355
BrewMethod      31557
SugarScale      31668
ABV             31654
FG              31690
IBU             31609
Style           33443
dtype: int64

In [28]:
li = lambda q1, ri: q1 - 1.5*ri
ls = lambda q3, ri: q3 + 1.5*ri
ri = lambda q1, q3: q3-q1

In [30]:
data_train['Efficiency'].describe()

count     314531
unique       355
top         70.0
freq       75643
Name: Efficiency, dtype: object

In [23]:
for col in data_train:
    print(f"{col} = {data_train[col].unique()}")

Efficiency = ['856842215413' '70.0' None '71' '80.0' '65.0' '75.0' '85.0' '59.0' '60.0'
 '77.0' '100.0' '72' '72.0' '75' '70' '71.0' '83.0' '35.0' '78.0'
 '-354128741541' '68.0' '73' '62.0' '82.0' '73.0' '40.0' '15.0' '66.0'
 '0.0' '25.0' '67.0' '10.0' '56' '87.0' '30' '65' '51.0' '50.0' '60'
 '55.0' '74.0' '63.0' '35' '76.0' '0.8' '66.3' '80.5' '45.0' '79.0' '77'
 '55' '80' '86.0' '69' '58.0' '83' '66' '30.0' '25' '76' '84.0' '73.9'
 '81.0' '90.0' '88.0' '61' '64.0' '78' '46.0' '95.0' '50' '51' '69.0'
 '89.0' '63' '92.0' '54.0' '63.5' '38.0' '100' '39.5' '85' '61.0' '67'
 '57.0' '53.0' '82' '52.0' 'fad' '23.0' '91.0' '83.8' '64' '49.0' '73.2'
 '72.5' '84' '90' '62' '54.5' '81' '61.5' '5.0' '43.0' '92' '65.7' '79'
 '58' '68.5' '40' '72.8' '73.5' '33' '42.0' '67.9' '1046.0' '93.0' '74'
 '83.9' '68' '53' '51.5' '77.5' '45' '65.5' '27.0' '93.9' '47.5' '33.6'
 '49' '37.0' '20.0' '56.0' '94.0' '44.0' '78.5' '71.3' '66.5' '84.3'
 '1044.0' '83.1' '0' '8.3' '49.3' '62.5' '61.6' '47.0' '57' '59